javascript/ql/src/semmle/javascript/security/UselessUseOfCat.qll

```diff
@@ -0,0 +1,339 @@
+   */
+  predicate isSync() { command.isSync() }
+
+  /**
+   * Gets an argument to this command execution that specifies the argument list to the command.
+   */
+  DataFlow::Node getArgumentList() { result = command.getArgumentList() }
+
+  /**
+   * Gets the callback (if it exists) for an async `exec`-like call.
+   */
+  DataFlow::FunctionNode getCallback() {
+    not this.isSync() and result = getLastArgument().getALocalSource()
+  }
+
+  /**
+   * Holds if the executed command execution has an argument list as a separate argument.
+   */
+  predicate hasArgumentList() { exists(command.getArgumentList()) }
+
+  /**
+   * Gets the data-flow node (if it exists) for a options argument for an `exec`-like call.
+   */
+  DataFlow::Node getOptionsArg() {
+    exists(int n |
+      n >= 1 and
+      // if there is a command-list, then the options is at least the third argument.
+      (not exists(command.getArgumentList()) or n >= 2) and
+      // async exec calls can have a callback as their last call.
+      if command.isSync() or not exists(getCallback())
+      then n < getNumArgument()
+      else n < getNumArgument() - 1
+    |
+      result = getArgument(n)
+    )
+    or
+    // Fallback in case normal API conventions are broken.
+    result = getAnArgument() and
+    result.getALocalSource() instanceof DataFlow::ObjectLiteralNode
+  }
+
+  /**
+   * Gets the constant-string parts that are not part of the command itself.
+   * E.g. for a command execution `exec("/bin/cat foo bar")` this predicate will have result `"foo bar"`.
+   */
+  string getNonCommandConstantString() {
+    if this.hasArgumentList()
+    then
+      result =
+        getConstantStringParts(getArgumentList()
+              .getALocalSource()
+              .(DataFlow::ArrayCreationNode)
+              .getElement(_))
+    else
+      exists(string commandString | commandString = getConstantStringParts(getArgument(0)) |
+        result = commandString.suffix(1 + commandString.indexOf(" ", 0, 0))
+      )
+  }
+
+  /**
+   * Holds if this command execution invokes the executeable `name`.
+   */
+  bindingset[name]
+  predicate isACallTo(string name) {
+    if this.hasArgumentList()
+    then getArgument(0).mayHaveStringValue(name)
+    else
+      exists(string arg | arg = getConstantStringParts(getArgument(0)) |
+        arg.prefix(name.length()) = name
+      )
+  }
+}
+
+/**
+ * Gets the constant string parts from a data-flow node.
+ * Either the string is some constant
+ */
+private string getConstantStringParts(DataFlow::Node node) {
+  node.mayHaveStringValue(result)
+  or
+  result = node.(StringOps::ConcatenationRoot).getConstantStringParts()
+}
+
+/**
+ * A call to a useless use of `cat`.
+ */
+class UselessCat extends CommandCall {
+  UselessCat() {
+    this = command and
+    isACallTo(getACatExecuteable()) and
+    // There is a file to read, and not just a pair of quotes.
+    (
+      not exists(PrettyPrintCatCall::createFileArgument(this))
```

Comment:
This is what I was worried about. Can we avoid letting the query results (excluding the message) depend on the pretty printing implementation?